File: client/common/protocol.go
package common

import (
	"bytes"
	"encoding/binary"
	"net"
	"strconv"
)

// Protocol Struct to handle communication protocol
type Protocol struct {
	conn net.Conn
	ID byte
}

// NewProtocol Initializes a new Protocol instance
func NewProtocol(conn net.Conn, ID string) (*Protocol, error) {
	idInt, err := strconv.Atoi(ID)
	if err != nil {
		return nil, err
	}
	return &Protocol{conn: conn, ID: byte(idInt)}, nil
}

// SendMessage Sends a message with a header containing the length of the data
func (p *Protocol) SendMessage(data []byte) error {
	allData := append([]byte{p.ID}, data...)
	log.Debugf("byte array: %v", allData)

	dataLength := uint32(len(allData))
	var lengthBuffer bytes.Buffer
	log.Debugf("length: %d", dataLength)

	err := binary.Write(&lengthBuffer, binary.BigEndian, dataLength)
	if err != nil {
		return err
	}

	if err := p.sendAll(lengthBuffer.Bytes()); err != nil {
		return err
	}

	return p.sendAll(allData)
}

// sendAll Ensures that all data is sent, handling short writes
func (p *Protocol) sendAll(data []byte) error {
	totalBytesSent := 0
	for totalBytesSent < len(data) {
		bytesSent, err := p.conn.Write(data[totalBytesSent:])
		if err != nil {
			return err
		}
		totalBytesSent += bytesSent
	}
	return nil
}

// ReceiveMessage Receives a message by first reading the length header and then the data
func (p *Protocol) ReceiveMessage() ([]byte, error) {
	lengthBuffer := make([]byte, 4)
	if err := p.readAll(lengthBuffer); err != nil {
		return nil, err
	}

	var dataLength uint32
	lengthReader := bytes.NewReader(lengthBuffer)
	err := binary.Read(lengthReader, binary.BigEndian, &dataLength)
	if err != nil {
		return nil, err
	}

	data := make([]byte, dataLength)
	if err := p.readAll(data); err != nil {
		return nil, err
	}

	return data, nil
}

// readAll Ensures that all data is read, handling short reads
func (p *Protocol) readAll(data []byte) error {
	totalBytesRead := 0
	for totalBytesRead < len(data) {
		bytesRead, err := p.conn.Read(data[totalBytesRead:])
		if err != nil {
			return err
		}
		totalBytesRead += bytesRead
	}
	return nil
}

File: server/common/protocol.py
class Protocol:
    def __init__(self, client_sock):
        self.client_sock = client_sock
        self.client_sock.setblocking(0)
        self.client_sock.settimeout(2)

    def send_message(self, message: bytes):
        """
        Send a message with a header containing the length of the message.
        This method ensures that the message is sent completely, handling short writes.
        """
        message_length = len(message).to_bytes(4, byteorder='big')
        self._send_all(message_length)
        self._send_all(message)
    
    def receive_message(self) -> bytes:
        """
        Receive a message by first reading the length header, then the message itself.
        This method ensures that the message is received completely, handling short reads.
        """
        # Read the length header (4 bytes)
        length_data = self._recvall(4)
        if not length_data:
            raise ConnectionError("Failed to receive message length.")
        
        message_length = int.from_bytes(length_data, byteorder='big')
        try:
            message = self._recvall(message_length)
        except TimeoutError:
            message = ''
        return message

    def _send_all(self, data: bytes):
        """
        Ensures that all data is sent, handling short writes.
        """
        total_sent = 0
        while total_sent < len(data):
            sent = self.client_sock.send(data[total_sent:])
            if sent == 0:
                raise ConnectionError("Socket connection broken.")
            total_sent += sent

    def _recvall(self, size: int) -> bytes:
        """
        Ensures that all data is received, handling short reads.
        """
        data = b''
        while len(data) < size:
            part = self.client_sock.recv(size - len(data))
            if part == b'':
                raise ConnectionError("Socket connection broken.")
            data += part
        return data
